File: quick_median/quickmedian.java
import java.util.Arrays;
import java.util.Random;
import java.lang.ArrayIndexOutOfBoundsException;
import java.lang.Comparable;

public class quickmedian {

  public static boolean DEBUG=false;

  /*
   * Generate array of unique random numbers
   */
  public static Integer[] genData(int size) {
    Integer[] data = new Integer[size];
    for (int i = 0; i < size; i++) {
      data[i] = i;
    }

    int index;
    for (int i = size - 1; i > 0; i--) {
      index = (int)(i * Math.random());
      int temp = data[index];
      data[index] = data[i];
      data[i] = temp;
    }
    return data;
  }

  public static void main(String[] args) {
    Integer[] data;
    try {
      data = genData(Integer.parseInt(args[0]));
    } catch (ArrayIndexOutOfBoundsException e) {
      System.err.println("Bad arguments!");
      return;
    }

    QuickMedian<Integer> qm = new QuickMedian<Integer>(data);
    if (DEBUG) {
      System.out.println("Generating random data array of size " + data.length + " with Fisher-Yates algorithm...");
      System.out.println("Before: " + qm);
    }

    long start = System.nanoTime();
    int median = qm.findMedian();
    if (DEBUG) {
      System.out.println("After median search:  " + qm);
    }
    long elapsed = System.nanoTime() - start;
    System.out.println("Median = " + median);

    System.out.println("Time ("+data.length+"):\t\t" + elapsed + " nsecs");
  }
}

class QuickMedian<T extends Comparable<? super T>> {
	private static boolean DEBUG=false;

  private T[] A;
  private Random rand;

  public QuickMedian(T[] A) {
    this.rand = new Random();
    this.A = A;
  }

  public T findMedian() {
    return findMedian(0, A.length-1, (A.length-1)/2);
  }

  public T findMedian(int l, int r, int order) {
    if (l == r) {
      return A[l];
    }
    if (DEBUG) System.out.println(this.toString());
    int pivot = randomizedPartition(l,r);
    if (DEBUG) System.out.println("l = " + l + ", r = " + r + ", pivot = " + pivot);
    if (order == pivot) {
      return A[pivot];
    } else
    if (order < pivot) {
      return findMedian(l,pivot-1,order);
    } else {
      return findMedian(pivot+1,r,order);
    }
  }

  private void swap(int i, int j) {
    T t = A[i];
    A[i] = A[j];
    A[j] = t;
  }

  private int randomizedPartition(int l, int r) {
		if (DEBUG) System.out.println("l="+l+",r="+r);
    int pivot = rand.nextInt(r-l+1) + l;
    if (DEBUG) System.out.println("pivot = " + pivot + ", A[pivot] = " + A[pivot]);
    swap(r,pivot);
    return Partition(l,r,pivot);
  }

  private int Partition(int l, int r, int pivot) {
    T t = A[r];
    int i = l, j = l-1;
    if (DEBUG) System.out.println("swap:l="+l+",r="+r+":" + this.toString());
    for ( ; i < r; i++) {
      if (A[i].compareTo(t) <= 0) {
        swap(i,++j);
        if (DEBUG) System.out.println("swap: ["+i+"<->"+j+"]:" + this.toString());
      }
    }
    swap(j+1,r);
    if (DEBUG) System.out.println("swap: return="+(j+1)+":" + this.toString());
    return j+1;
  }
  @Override
  public String toString() {
    return new String(Arrays.toString(this.A));
  }
}

File: quick_median/quickmedian.cpp
/*
 * =====================================================================================
 *
 *       Filename:  quickmedian.cpp
 *
 *    Description:  Quick median (C++)
 *
 *        Version:  1.0
 *        Created:  02/14/2014 04:44:03 PM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Michael Likholet <m.likholet@ya.ru>
 *        Company:  -
 *
 * =====================================================================================
 */

#include <vector>
#include <iterator>
#include <iostream>
#include <string>
#include <algorithm>
#include <utility>
#include <sstream>

#include <cstdlib>

class QuickMedian {
	public:
		QuickMedian(std::vector<int> vec): data(vec) {}
		~QuickMedian() {}

		int findMedian();
		int findMedian(int l, int r, int order);

	private:
		std::vector<int> data;

		int Partition(int l, int r);
		int randomizedPartition(int l, int r);
};

int QuickMedian::findMedian() {
	return findMedian(0, data.size(), data.size()/2);
}

int QuickMedian::findMedian(int l, int r, int order) {
	if (l == r) {
    return data[l];
  }
  int pivot = randomizedPartition(l, r);
  if (order == pivot) {
    return data[order];
  } else 
  if (order < pivot) {
    return findMedian(l, pivot - 1, order);
  } else { /* if (order > pivot) */
    return findMedian(pivot + 1, r, order);
  }
}

int QuickMedian::Partition(int l, int r) {
  int t = data[r], i = l, j = l;
  for ( ; i < r; ++i) {
    if (data[i] < t) {
      std::swap(&data[i], &data[j++]);
    }
  }
  std::swap(&data[j], &data[r]);
  return j;
}

int QuickMedian::randomizedPartition(int l, int r) {
  int pivot = std::rand() % (r - l + 1) + l;
  swap(pivot, r);
  return Partition(l,r);
}

namespace FisherYatsShuffle {
  static void shuffle(vector<int> &input) {
    std::vector<int>::iterator it = input.rbegin();
    int idx = 0;
    for (it != input.rend(); it++) {
      idx = std::rand() % (it - input.begin());
      std::iter_swap(it,input.begin() + idx);
    }
  }
}

int main(int argc, char *argv[]) {

  std::chrono::time_point<std::chrono::system_clock> start, end;
  vector<int> *data = null;

	switch (argc) {
		case 2:
			std::cout << "Generating data..." << std::endl;
      istringstream ss(argv[1]);
      int length;
      if (!(ss >> length)) {
        std::cerr << "Bad argument!" << std::endl;
        exit(1);
      }
      start = std::chrono::system_clock::now();
      data = new std::vector<int>(length);
      for (std::vector<int>::iterator it = data.begin(); it != data.end(); ++it) {
        *it = it - data.begin();
      }
      FisherYatsShuffle::shuffle(data);
      end = std::chrono::system_clock::now();
      std::chrono::duration<double> elapsed_seconds = end - start; 
			std::cout << "Generating took " << elapsed_seconds.count() << " secs" << std::endl;
			break;
		default:
      std::cerr << "Bad arguments count!" << std::endl;
			exit(2);
	};

	std::cout << "Median = " << median << std::endl;
	std::cout << "Time = " << time << " usecs" << std::endl;

	start = std::chrono::system_clock::now();

  QuickMedian qm(data);
  int median = qm.findMedian();
  end = std::chrono::system_clock::now();
  std::cout << "Median = " << median << std::endl;
  std::chrono::duration<double> elapsed_seconds = end - start; 
  std::cout << "Computation took " << elapsed_seconds.count() << " secs" << std::endl;

	return 0;
}
